### PR TITLE
IP scrubber

### PR DESCRIPTION
**Pull request by VeronikaSolovei9**
Added `anonymizeIpv6` function to transmitPreciseGeo activitiy to match PBS-Java implementation

Addressed issue https://github.com/prebid/prebid-server/issues/2981

### CODE REVIEW

File: privacy/scrubber.go [hunk 1]
```diff
@@ -260,6 +262,12 @@ func scrubIPV6Lowest16Bits(ip string) string {
 	return ip
 }
 
+func anonymizeIpv6(ip string) string {
+	ipv6Mask := net.CIDRMask(56, 128)
+	ipMasked := net.ParseIP(ip).Mask(ipv6Mask)
+	return ipMasked.String()
+}
```
Comment (VeronikaSolovei9) [inline privacy/scrubber.go:269]: This function can be reused in other scrubbing strategies like 
```go
      // ScrubStrategyIPV6Lowest16 zeroes out the last 16 bits of an IPV6 address.
	ScrubStrategyIPV6Lowest16

	// ScrubStrategyIPV6Lowest32 zeroes out the last 32 bits of an IPV6 address.
	ScrubStrategyIPV6Lowest32
```
by passing number of bits that should remain un-anonymous to the first parameter here:`net.CIDRMask(56, 128)`

Comment (SyntaxNode) [reply privacy/scrubber.go:269]: Please change the name to `scrubIPV6` to match PBS-Go naming. This functionality should also replace `ScrubStrategyIPV6Lowest16` and `ScrubStrategyIPV6Lowest32` instead of exist beside it.

File: privacy/scrubber_test.go [hunk 2]
```diff
@@ -586,6 +586,37 @@ func TestScrubIPV6Lowest32Bits(t *testing.T) {
 	}
 }
 
+func TestAnonymizeIpv6(t *testing.T) {
+	testCases := []struct {
+		IP          string
+		cleanedIP   string
+		description string
```
Comment (SyntaxNode) [inline privacy/scrubber_test.go:593]: Please change `description` to `name` to match the arguments of `t.Run` and provide a simple and short test name per case. These tests do not have a `description`.

Alternatively for this test, using the IP as the test name might work well.

File: privacy/scrubber.go [hunk 3]
```diff
@@ -260,6 +262,12 @@ func scrubIPV6Lowest16Bits(ip string) string {
 	return ip
 }
 
+func anonymizeIpv6(ip string) string {
+	ipv6Mask := net.CIDRMask(56, 128)
```
Comment (SyntaxNode) [inline privacy/scrubber.go:266]: To match PBS-Java, these two variables should be read from the config (account config?) with appropriate validation on the values.

Comment (VeronikaSolovei9) [reply privacy/scrubber.go]: In progress, not yet ready for review

File: config/account.go [hunk 4]
```diff
@@ -297,3 +303,36 @@ func (a *AccountChannel) IsSet() bool {
...
+	if ipMasks.GdprLeftMaskBitsLowest > maxBits || ipMasks.GdprLeftMaskBitsLowest < 0 {
+		err := fmt.Errorf("gdpr left mask bits lowest cannot exceed %d in %s address, or be less than 0", maxBits, ipVersion)
+		errs = append(errs, err)
+	}
+	if ipMasks.GdprLeftMaskBitsHighest > maxBits || ipMasks.GdprLeftMaskBitsHighest < 0 {
+		err := fmt.Errorf("gdpr left mask bits highest cannot exceed %d in %s address, or be less than 0", maxBits, ipVersion)
+		errs = append(errs, err)
+	}
+	return errs
+}
```
Comment (SyntaxNode) [inline config/account.go:338]: I recommend to break the errors array pattern here. It forces each validate method to understand too much context from it's caller. Instead, consider returning an array and appending as necessary. Something like:

```
func (ipMasks *IpMasks) validate(maxBits int, ipVersion string) []error {
	var errs []error

	if ipMasks.ActivityLeftMaskBits > maxBits || ipMasks.ActivityLeftMaskBits < 0 {
		err := fmt.Errorf("activity left mask bits cannot exceed %d in %s address, or be less than 0", maxBits, ipVersion)
		errs = append(errs, err)
	}

	if ipMasks.GdprLeftMaskBitsLowest > maxBits || ipMasks.GdprLeftMaskBitsLowest < 0 {
		err := fmt.Errorf("gdpr left mask bits lowest cannot exceed %d in %s address, or be less than 0", maxBits, ipVersion)
		errs = append(errs, err)
	}

	if ipMasks.GdprLeftMaskBitsHighest > maxBits || ipMasks.GdprLeftMaskBitsHighest < 0 {
		err := fmt.Errorf("gdpr left mask bits highest cannot exceed %d in %s address, or be less than 0", maxBits, ipVersion)
		errs = append(errs, err)
	}

	return errs
}

func (ipMasking *IpMasking) Validate(errs []error) []error {
        var errs []error

	if e := ipMasking.IpV6.validate(errs, Ipv6Bits, "ipv6"); e != nil {
                errs = append(errs, e)
        }
        
	if errs = ipMasking.IpV4.validate(errs, Ipv4Bits, "ipv4") ; e != nil {
                errs = append(errs, e)
        }

	return errs
}
```

File: config/account.go [hunk 5]
```diff
@@ -297,3 +303,36 @@ func (a *AccountChannel) IsSet() bool {
...
+
+type IpMasks struct {
+	ActivityLeftMaskBits    int `mapstructure:"activity_left_mask_bits" json:"activity_left_mask_bits"`
+	GdprLeftMaskBitsLowest  int `mapstructure:"gdpr_left_mask_bits_lowest" json:"gdpr_left_mask_bits_lowest"`
+	GdprLeftMaskBitsHighest int `mapstructure:"gdpr_left_mask_bits_highest" json:"gdpr_left_mask_bits_highest"`
+}
+
+func (ipMasking *IpMasking) Validate(errs []error) []error {
+	errs = ipMasking.IpV6.validate(errs, Ipv6Bits, "ipv6")
```
Comment (SyntaxNode) [inline config/account.go:319]: You need to handle the case of `ipMasking` being nil.

File: privacy/enforcement.go [hunk 6]
```diff
@@ -27,8 +30,8 @@ func (e Enforcement) AnyActivities() bool {
 }
 
 // Apply cleans personally identifiable information from an OpenRTB bid request.
-func (e Enforcement) Apply(bidRequest *openrtb2.BidRequest) {
-	e.apply(bidRequest, NewScrubber())
+func (e Enforcement) Apply(bidRequest *openrtb2.BidRequest, ipMasking *config.IpMasking) {
```
Comment (SyntaxNode) [inline privacy/enforcement.go:33]: Does this need to be passed as a pointer?

File: privacy/scrubber.go [hunk 7]
```diff
@@ -76,14 +77,18 @@ type Scrubber interface {
 	ScrubUser(user *openrtb2.User, strategy ScrubStrategyUser, geo ScrubStrategyGeo) *openrtb2.User
 }
 
-type scrubber struct{}
+type scrubber struct {
+	ipMasking *config.IpMasking
```
Comment (SyntaxNode) [inline privacy/scrubber.go:81]: Does this need to be passed as a pointer?

File: privacy/scrubber.go [hunk 8]
```diff
@@ -176,7 +181,7 @@ func (scrubber) ScrubRequest(bidRequest *openrtb2.BidRequest, enforcement Enforc
 	return bidRequest
 }
 
-func (scrubber) ScrubDevice(device *openrtb2.Device, id ScrubStrategyDeviceID, ipv4 ScrubStrategyIPV4, ipv6 ScrubStrategyIPV6, geo ScrubStrategyGeo) *openrtb2.Device {
+func (s *scrubber) ScrubDevice(device *openrtb2.Device, id ScrubStrategyDeviceID, ipv4 ScrubStrategyIPV4, ipv6 ScrubStrategyIPV6, geo ScrubStrategyGeo) *openrtb2.Device {
```
Comment (SyntaxNode) [inline privacy/scrubber.go:184]: Why the change to a pointer receiver?

File: privacy/scrubber.go [hunk 9]
```diff
@@ -196,14 +201,14 @@ func (scrubber) ScrubDevice(device *openrtb2.Device, id ScrubStrategyDeviceID, i
 
 	switch ipv4 {
 	case ScrubStrategyIPV4Lowest8:
-		deviceCopy.IP = scrubIPV4Lowest8(device.IP)
+		deviceCopy.IP = scrubIp(device.IP, s.ipMasking.IpV4.GdprLeftMaskBitsLowest, config.Ipv4Bits)
```
Comment (SyntaxNode) [inline privacy/scrubber.go:204]: Instead of needing to reference such a deep property, consider storing IPv4 and IPv6 in the scrubber struct.

File: privacy/scrubber.go [hunk 10]
```diff
@@ -241,44 +246,13 @@ func (scrubber) ScrubUser(user *openrtb2.User, strategy ScrubStrategyUser, geo S
 	return &userCopy
 }
 
-func scrubIPV4Lowest8(ip string) string {
-	i := strings.LastIndex(ip, ".")
-	if i == -1 {
+func scrubIp(ip string, ones, bits int) string {
```
Comment (SyntaxNode) [inline privacy/scrubber.go:249]: Nitpick: `scrubIP`

File: privacy/scrubber.go (hunk 7 above)
Comment (VeronikaSolovei9) [reply privacy/scrubber.go]: My main idea here was to load these values to a non-pointer config struct, because we always need them. And then pass it as a pointer downstream to avoid copying the struct. 
But I can see not why this was not a good idea. I replaced it to non-pointer.

File: privacy/scrubber.go (hunk 8 above)
Comment (VeronikaSolovei9) [reply privacy/scrubber.go]: Revert back

File: privacy/enforcement.go (hunk 6 above)
Comment (VeronikaSolovei9) [reply privacy/enforcement.go]: Modified

File: config/account.go (hunk 5 above)
Comment (VeronikaSolovei9) [reply config/account.go]: Technically `ipMasking` is never nil here. Should I modify it to a non-pointer?

File: config/account.go (hunk 4 above)
Comment (VeronikaSolovei9) [reply config/account.go:347]: Modified

File: config/account.go [hunk 11]
```diff
@@ -297,3 +303,36 @@ func (a *AccountChannel) IsSet() bool {
 type AccountPrivacy struct {
 	AllowActivities AllowActivities `mapstructure:"allowactivities" json:"allowactivities"`
 }
+
+type IpMasking struct {
```
[the remark this answers is not in the document]
Comment (VeronikaSolovei9) [reply config/account.go]: Modified

File: config/account.go [hunk 12]
```diff
@@ -21,6 +21,11 @@ const (
 	ChannelWeb   ChannelType = "web"
 )
 
+const (
+	Ipv4Bits = 32
```
[the remark this answers is not in the document]
Comment (VeronikaSolovei9) [reply config/account.go]: Renamed

File: privacy/scrubber.go (hunk 3 above)
Comment (VeronikaSolovei9) [reply privacy/scrubber.go]: Moved to `account.privacy`

File: privacy/scrubber.go (hunk 7 above)
Comment (SyntaxNode) [reply privacy/scrubber.go]: > And then pass it as a pointer downstream to avoid copying the struct.

That is't a surefire premature optimization. Go has optimizations for passing arguments as values which may be much more efficient than using a pointer. It's a bit more reliable for large data structures like the BidRequest, but this struct is relatively small.

File: config/account.go [hunk 13]
```diff
@@ -21,6 +21,11 @@ const (
 	ChannelWeb   ChannelType = "web"
 )
 
+const (
+	IPv4 = 32
+	IPv6 = 128
```
Comment (SyntaxNode) [inline config/account.go:26]: Looking at this again, you may wish to consider adding more information to describe what the numbers mean. Perhaps such as `IPv4Bits` / `IPv4BitSize` / `IPv4BitLength`?

I also missed that if you don't want to export these constants, the names `ipv4` and `ipv6` are equally good choices. Just so long as the `I` and `P` match case.

File: config/account.go [hunk 14]
```diff
@@ -295,5 +300,47 @@ func (a *AccountChannel) IsSet() bool {
 }
 
 type AccountPrivacy struct {
-	AllowActivities AllowActivities `mapstructure:"allowactivities" json:"allowactivities"`
+	AllowActivities *AllowActivities `mapstructure:"allowactivities" json:"allowactivities"`
+	IPMasking       IPMasking        `mapstructure:"ip_masking" json:"ip_masking"`
+}
+
+type IPMasking struct {
+	IpV6 IPMasks `mapstructure:"ipv6" json:"ipv6"`
```
Comment (SyntaxNode) [inline config/account.go:308]: Variable naming "IPv6" / "IPv4"

File: config/account.go [hunk 15]
```diff
@@ -295,5 +300,47 @@ func (a *AccountChannel) IsSet() bool {
...
+	IpV4 IPMasks `mapstructure:"ipv4" json:"ipv4"`
+}
+
+type IPMasks struct {
+	ActivityLeftMaskBits    int `mapstructure:"activity_left_mask_bits" json:"activity_left_mask_bits"`
+	GdprLeftMaskBitsLowest  int `mapstructure:"gdpr_left_mask_bits_lowest" json:"gdpr_left_mask_bits_lowest"`
+	GdprLeftMaskBitsHighest int `mapstructure:"gdpr_left_mask_bits_highest" json:"gdpr_left_mask_bits_highest"`
+}
```
Comment (SyntaxNode) [inline config/account.go:316]: Comparing to PBS-Java, they don't have `activity_left_mask_bits` and instead have just `always-mask-right` and `anon-left-mask-bits`. Why aren't we matching that?

Comment (VeronikaSolovei9) [reply config/account.go:316]: Yes, this is something that we have different.
`always-mask-right` (default value is 64):
They always cut 64 right bits, which is a device MAC address or device identifier.
(For visualization) it turns 
`1111:2222:3333:4444:5555:6666:7777:8888`
to 
`1111:2222:3333:4444::`

`anon-left-mask-bits` (default value is 56)
Is used in activity IPv6 scrubbing. 
It turns 
`1111:2222:3333:4444:5555:6666:7777:8888`
to 
`1111:2222:3333:4400::`

In PBS we have IPv6 configs for activities and 2 more configs for GDPR scrubbing strategies: 
```go
// ScrubStrategyIPV6Lowest16 zeroes out the last 16 bits of an IPV6 address.
ScrubStrategyIPV6Lowest16

// ScrubStrategyIPV6Lowest32 zeroes out the last 32 bits of an IPV6 address.
ScrubStrategyIPV6Lowest32
```
This is why I added `gdpr_left_mask_bits_lowest` and `gdpr_left_mask_bits_highest`.  I didn't find corresponding code in PBS-Java.

I also think lowest 16 or lowest 32 should be scrubbed from network address, not from device identifier. I might be wrong about this.

Current PBS `scrubIPV6Lowest16Bits` turns 
`1111:2222:3333:4444:5555:6666:7777:8888`
to 
`1111:2222:3333:4444:5555:6666:7777:0`

current PBS `scrubIPV6Lowest32Bits` turns 
`1111:2222:3333:4444:5555:6666:7777:8888`
to 
`1111:2222:3333:4444:5555:6666:0:0`

Instead it should be 
`scrubIPV6Lowest16Bits`: `1111:2222:3333::`
`scrubIPV6Lowest32Bits`: `1111:2222::`

These both values are easy to change in configs now. We just need to figure out what it should be.

File: config/account.go (hunk 13 above)
Comment (VeronikaSolovei9) [reply config/account.go:26]: Renamed to IPv4BitSize and IPv6BitSize. Yes, they are exported.

File: config/account.go (hunk 15 above)
Comment (VeronikaSolovei9) [reply config/account.go:316]: Here is a possible config change:
```go
v.SetDefault("account_defaults.privacy.ip_masking.ipv6.activity_left_mask_bits", 56)
v.SetDefault("account_defaults.privacy.ip_masking.ipv6.gdpr_left_mask_bits_lowest", 48) // now 112 (112-64=48)
v.SetDefault("account_defaults.privacy.ip_masking.ipv6.gdpr_left_mask_bits_highest", 32) // now 96 (96-64=32)
```

Comment (VeronikaSolovei9) [reply config/account.go:314]: Discussed offline. Only `left_mask_bits` variable should be present in ip masking config.

File: account/account.go [hunk 16]
```diff
@@ -103,6 +103,11 @@ func GetAccount(ctx context.Context, cfg *config.Configuration, fetcher stored_r
 		return nil, errs
 	}
 
+	errs = account.Privacy.IPMasking.Validate(errs)
+	if len(errs) > 0 {
+		return nil, errs
```
Comment (AlexBVolcy) [inline account/account.go:108]: Is there a reasonable way to add a test that throws this error? Or is it challenging with this Validate function?

File: config/account_test.go [hunk 17]
```diff
@@ -910,3 +910,61 @@ func TestAccountPriceFloorsValidate(t *testing.T) {
 		})
 	}
 }
+
+func TestIPMaskingValidate(t *testing.T) {
+
```
Comment (AlexBVolcy) [inline config/account_test.go:915]: Nitpick: Could you remove this blank line?

File: config/account.go [hunk 18]
```diff
@@ -40,7 +45,7 @@ type Account struct {
 	Validations             Validations                                 `mapstructure:"validations" json:"validations"`
 	DefaultBidLimit         int                                         `mapstructure:"default_bid_limit" json:"default_bid_limit"`
 	BidAdjustments          *openrtb_ext.ExtRequestPrebidBidAdjustments `mapstructure:"bidadjustments" json:"bidadjustments"`
-	Privacy                 *AccountPrivacy                             `mapstructure:"privacy" json:"privacy"`
+	Privacy                 AccountPrivacy                              `mapstructure:"privacy" json:"privacy"`
```
Comment (AlexBVolcy) [inline config/account.go:48]: For my understanding, why is this converted to remove its pointer?

Comment (VeronikaSolovei9) [reply config/account.go:48]: I changed it to non-pointer because we no have a new `IPMasking` field inside of the Privacy: 
```go
type AccountPrivacy struct {
	AllowActivities *AllowActivities `mapstructure:"allowactivities" json:"allowactivities"`
	IPMasking       IPMasking        `mapstructure:"ip_masking" json:"ip_masking"` 
}
```
We need to have IPMasking to be always present and to be populated with default values from configs. 
PBS host account_defaults configs and/or request.account config can overwrite these values. 

This is why config.config.go has these lines: 
```go
v.SetDefault("account_defaults.privacy.ip_masking.ipv6.left_mask_bits", 56)
v.SetDefault("account_defaults.privacy.ip_masking.ipv4.left_mask_bits", 24)
```

File: account/account.go (hunk 16 above)
Comment (VeronikaSolovei9) [reply account/account.go:108]: Yes, added!

File: privacy/scrubber_test.go [hunk 19]
```diff
@@ -462,127 +448,110 @@ func TestScrubRequest(t *testing.T) {
...
 
-			result := NewScrubber().ScrubRequest(bidRequest, test.enforcement)
+			result := NewScrubber(testIPMasking).ScrubRequest(bidRequest, test.enforcement)
 			assert.Equal(t, test.expected, result, test.description)
 		})
 	}
 }
 
-func TestScrubIPV4(t *testing.T) {
+func TestScrubIp(t *testing.T) {
```
Comment (AlexBVolcy) [inline privacy/scrubber_test.go:457]: Nitpick: Could you capitalize the `Ip` here so it looks like `TestScrubIP`

**Review by AlexBVolcy — COMMENTED** [review]
Looks good! Just left one little nitpick

**Review by AlexBVolcy — DISMISSED** [review]
LGTM

File: config/account.go (hunk 15 above)
Comment (SyntaxNode) [reply config/account.go:314]: Why are not using the same  `always-mask-right` and `anon-left-mask-bits` approach that PBS-Java has? I like the names you've been using better than though, since "mask-right" doesn't look like it ever makes actually masks the right side?

Comment (VeronikaSolovei9) [reply config/account.go:314]: In Java there 2 variables used in IPv6 processing: 
```yaml
ipv6:
  always-mask-right: 64
  anon-left-mask-bits: 56
```
If I understand the Java code correctly, they always remove 64 bits from the right for any incoming IPv6 address, even if GDPR is not applied:
`1111:2222:3333:4444:5555:6666:7777:8888` -> `1111:2222:3333:4444::`

If GDPR is applied and IPv6 should be anonymized only 56 bits from the left will remain. This is what we have in this PR. We only modify IPv6 if GDPR or activity is enforced: 
`1111:2222:3333:4444:5555:6666:7777:8888` -> `1111:2222:3333:4400::`

Do we also want to remove 64 bits from the right even if GDPR was not applied?

File: config/account.go [hunk 20]
```diff
@@ -301,35 +301,32 @@ func (a *AccountChannel) IsSet() bool {
...
+	IPv4Config      IPv4             `mapstructure:"ipv4" json:"ipv4"`
 }
 
-type IPMasking struct {
-	IPv6 IPMasks `mapstructure:"ipv6" json:"ipv6"`
-	IPv4 IPMasks `mapstructure:"ipv4" json:"ipv4"`
+type IPv6 struct {
+	AnonKeepBits int `mapstructure:"anon-keep-bits" json:"anon-keep-bits"`
+	// For the follow up PR, always keep first 64 bits
+	// AlwaysKeepBits int `mapstructure:"always-keep-bits" json:"always-keep-bits"`
```
Comment (AlexBVolcy) [inline config/account.go:311]: Just want to confirm that we want to keep this commented code out here?

**Review by AlexBVolcy — COMMENTED** [review]
Left one comment. Could you also resolve the merge conflicts?

File: account/account.go [hunk 21]
```diff
@@ -103,6 +103,16 @@ func GetAccount(ctx context.Context, cfg *config.Configuration, fetcher stored_r
 		return nil, errs
 	}
 
+	errs = account.Privacy.IPv6Config.Validate(errs)
+	if len(errs) > 0 {
+		return nil, errs
+	}
```
Comment (SyntaxNode) [inline account/account.go:109]: This check may be controversial. The preferred approach we now take to account config error handling is to record a metric and revert to defaults such that auctions can still run using as much of the account config that's valid. The validation may be better suited for the `/ accountID resolved to a valid account, merge with AccountDefaults for a complete config` branch.

I assume we don't need to validate the default account because that's handled separatly on app start?

File: privacy/scrubber.go [hunk 22]
```diff
@@ -199,14 +203,12 @@ func (scrubber) ScrubDevice(device *openrtb2.Device, id ScrubStrategyDeviceID, i
 
 	switch ipv4 {
 	case ScrubStrategyIPV4Lowest8:
```
Comment (SyntaxNode) [inline privacy/scrubber.go:205]: Why do we still need `ScrubStrategyIPV4Lowest8`?

File: config/account.go [hunk 23]
```diff
@@ -295,5 +300,33 @@ func (a *AccountChannel) IsSet() bool {
...
+}
+
+type IPv4 struct {
+	AnonKeepBits int `mapstructure:"anon-keep-bits" json:"anon-keep-bits"`
+}
+
+func (IPv6Config *IPv6) Validate(errs []error) []error {
+	if IPv6Config.AnonKeepBits > IPv6BitSize || IPv6Config.AnonKeepBits < 0 {
```
[the remark this answers is not in the document]
Comment (VeronikaSolovei9) [reply config/account.go:319]: IPv6 and IPv4 are not pointers in AccountPrivacy:
```go
type AccountPrivacy struct {
	AllowActivities *AllowActivities `mapstructure:"allowactivities" json:"allowactivities"`
	IPv6Config      IPv6             `mapstructure:"ipv6" json:"ipv6"`
	IPv4Config      IPv4             `mapstructure:"ipv4" json:"ipv4"`
}
```
technically they will be always initialized. Should I add nil check anyways?

[the remark this answers is not in the document]
Comment (VeronikaSolovei9) [reply config/account.go:319]: IPv6 and IPv4 are not pointers in AccountPrivacy:
```go
type AccountPrivacy struct {
	AllowActivities *AllowActivities `mapstructure:"allowactivities" json:"allowactivities"`
	IPv6Config      IPv6             `mapstructure:"ipv6" json:"ipv6"`
	IPv4Config      IPv4             `mapstructure:"ipv4" json:"ipv4"`
}
```
technically they will be always initialized. Should I add a nil check anyways?

File: privacy/scrubber_test.go [hunk 24]
```diff
@@ -38,12 +39,12 @@ func TestScrubDevice(t *testing.T) {
 				MACMD5:   "",
 				IFA:      "",
 				IP:       "1.2.3.0",
-				IPv6:     "2001:0db8:0000:0000:0000:ff00:0:0",
+				IPv6:     "2001:db8:2233:4400::",
```
[the remark this answers is not in the document]
Comment (VeronikaSolovei9) [reply privacy/scrubber_test.go:42]: In this case `0db8` and `db8` are identical. Because we use `net.ParseIP(ip)` and then `ipMasked.String()` returns it in this form.

[the remark this answers is not in the document]
Comment (VeronikaSolovei9) [reply privacy/scrubber_test.go:42]: https://www.ciscopress.com/articles/article.asp?p=2803866 -> paragraph "Rule 1: Omit Leading 0s"

File: privacy/scrubber.go (hunk 22 above)
Comment (VeronikaSolovei9) [reply privacy/scrubber.go:205]: We need this to indicate we want to scrun IPv4. There are just 2 options:
```go
const (
	// ScrubStrategyIPV4None does not remove any part of an IPV4 address.
	ScrubStrategyIPV4None ScrubStrategyIPV4 = iota

	// ScrubStrategyIPV4Lowest8 zeroes out the last 8 bits of an IPV4 address.
	ScrubStrategyIPV4Lowest8
)
```
I think we should rename it to something like `ScrubStrategyIPV4Default` or better.

File: account/account.go (hunk 21 above)
Comment (VeronikaSolovei9) [reply account/account.go:109]: Is it ok if metrics will be added in a separate PR? I modified account validation and in case data is not valid we set it to default.

Default config can be overwritten by host config. I assume we still need a host config ipV4 and ipV6 validation. I'm ok to remove host config validation

File: account/account.go [hunk 25]
```diff
@@ -103,14 +103,18 @@ func GetAccount(ctx context.Context, cfg *config.Configuration, fetcher stored_r
...
 	}
 
-	errs = account.Privacy.IPv4Config.Validate(errs)
-	if len(errs) > 0 {
-		return nil, errs
+	var ipv4Errs []error
+	ipv4Errs = account.Privacy.IPv4Config.Validate(ipv4Errs)
+	if len(ipv4Errs) > 0 {
+		account.Privacy.IPv4Config.AnonKeepBits = iputil.IPv4BitSize
+		// record invalid ipv4 metric
```
Comment (AlexBVolcy) [inline account/account.go:117]: Same point about removing metric related comment

File: config/account.go (hunk 5 above)
Comment (SyntaxNode) [reply config/account.go]: Yes, please. Let's add some defensive coding since `nil` is allowed here. Alternatively, you may be able to change this to a value type receiver.

File: privacy/scrubber_test.go (hunk 24 above)
[the remark this answers is not in the document]
Comment (SyntaxNode) [reply privacy/scrubber_test.go:42]: Gotcha. Maybe just add a the `0` to keep this straight forward?

File: account/account.go (hunk 21 above)
Comment (SyntaxNode) [reply account/account.go:111]: Absolutely.

File: privacy/scrubber.go (hunk 22 above)
Comment (SyntaxNode) [reply privacy/scrubber.go:206]: Ah ok. Yes, renaming would be good. Perhaps `ScrubStrategyIPV4Subnet` to match IPv6 naming?

Comment (VeronikaSolovei9) [reply privacy/scrubber.go:206]: Right!

File: privacy/scrubber_test.go (hunk 24 above)
[the remark this answers is not in the document]
Comment (VeronikaSolovei9) [reply privacy/scrubber_test.go:42]: Ooof, I see. This value is returned by the go `net.IP` functionality. In order to make this change we will need to compare net.IPs, not strings. This will require some assertions refactoring. We now assert the whole `expected: &openrtb2.Device` with the result.

We will need to change this(current implementation): 
```go
assert.Equal(t, test.expected, result, test.description)
```
to something like this:
```go
assert.Equal(t, net.ParseIP(test.expected.IP).String(), net.ParseIP(result.IP).String(), test.description)
assert.Equal(t, net.ParseIP(test.expected.IPv6).String(), net.ParseIP(result.IPv6).String(), test.description)
assert.Equal(t, test.expected.DIDMD5, result.DIDMD5, test.description)
assert.Equal(t, test.expected.IFA, result.IFA, test.description)
...
```

 Are we good to go with this format?

[the remark this answers is not in the document]
Comment (SyntaxNode) [reply privacy/scrubber_test.go:42]: How about we just change 0 to 1? My only intention here was to keep the test straight forward and not confuse the reader with the additional dimension of leading 0 optimization.

File: account/account.go [hunk 26]
```diff
@@ -103,6 +103,18 @@ func GetAccount(ctx context.Context, cfg *config.Configuration, fetcher stored_r
...
+	ipv6Errs = account.Privacy.IPv6Config.Validate(ipv6Errs)
+	if len(ipv6Errs) > 0 {
+		account.Privacy.IPv6Config.AnonKeepBits = iputil.IPv6BitSize
+	}
+
+	var ipv4Errs []error
+	ipv4Errs = account.Privacy.IPv4Config.Validate(ipv4Errs)
+	if len(ipv4Errs) > 0 {
+		account.Privacy.IPv4Config.AnonKeepBits = iputil.IPv4BitSize
+	}
```
Comment (SyntaxNode) [inline account/account.go:116]: Nitpick: Could simplify the syntax...

```
if errs := account.Privacy.IPv6Config.Validate(nil); len(errs) > 0 {
	account.Privacy.IPv6Config.AnonKeepBits = iputil.IPv6BitSize
}

if errs := account.Privacy.IPv4Config.Validate(nil); len(errs) > 0 {
	account.Privacy.IPv4Config.AnonKeepBits = iputil.IPv4BitSize
}
```

Comment (SyntaxNode) [reply account/account.go:116]: These lines show as having test coverage, but there doesn't seem to an assertion on setting their default values. I made a type locally where `IPv4Config.AnonKeepBits = iputil.IPv6BitSize` which is setting the ipv6 value for the ipv4 default and this mistake wasn't caught by a test.

File: config/account.go [hunk 27]
```diff
@@ -295,5 +296,31 @@ func (a *AccountChannel) IsSet() bool {
...
+	IPv4Config      IPv4             `mapstructure:"ipv4" json:"ipv4"`
+}
+
+type IPv6 struct {
+	AnonKeepBits int `mapstructure:"anon-keep-bits" json:"anon-keep-bits"`
+}
+
+type IPv4 struct {
+	AnonKeepBits int `mapstructure:"anon-keep-bits" json:"anon-keep-bits"`
+}
```
Comment (SyntaxNode) [inline config/account.go:310]: Nitpick: Let's use underscores to be consistent with the rest of this object: `anon-keep-bits` -> `anon_keep_bits`

File: privacy/scrubber_test.go (hunk 24 above)
[the remark this answers is not in the document]
Comment (SyntaxNode) [reply privacy/scrubber_test.go:42]: How about we change to 0 to a 1 to remove this issue entirely?

File: account/account.go (hunk 26 above)
Comment (VeronikaSolovei9) [reply account/account.go:112]: Yes, this is tested in a general GetAccount test. 
I added extra flag to assert IP config for account id "invalid_acct_ipv6_ipv4". 
I also realized I set both ipv6 and ipv4 to max value, instead of default mask bits. I added 2 new constants: 
```
IPv4DefaultMaskingBitSize = 24
IPv6DefaultMaskingBitSize = 56
```
For the simplified syntax I added new variables for errors to show explicitly we don't want to add IP config validation errors to `errs`.

Comment (VeronikaSolovei9) [reply account/account.go:112]: And also tested this update end to end including test with publisher account id.

**Review by SyntaxNode — APPROVED** [review]
